### PR TITLE
[SPARK-21291][R] add R partitionBy API in DataFrame

### PR DESCRIPTION
**Pull request by huaxingao**
## What changes were proposed in this pull request?

add R partitionBy API in write.df
I didn't add bucketBy in write.df. The last line of write.df is
```
write <- handledCallJMethod(write, "save")
```
save doesn't support bucketBy right now. 
```
 assertNotBucketed("save")
```

## How was this patch tested?

Add unit test in test_sparkSQL.R

### CODE REVIEW

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by felixcheung** [comment]
could you kick off appveyor again by close/re-open this PR?
also might need to split the JIRA - since it's referring to both partition and bucket

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -2985,8 +2988,18 @@ setMethod("write.df",
             if (is.null(source)) {
               source <- getDefaultSqlSource()
             }
+            cols <- NULL
+            if (!is.null(partitionBy)) {
+              if (!all(sapply(partitionBy, function(c) { is.character(c) }))) {
```
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R]: not a big deal but `function(c) { is.character(c) }` -> `function(c) is.character(c)` if more changes should be pushed.

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 2]
```diff
@@ -2713,8 +2713,16 @@ test_that("read/write text files", {
   expect_equal(colnames(df2), c("value"))
   expect_equal(count(df2), count(df) * 2)
 
+  df3 <- createDataFrame(list(list(1L, "1"), list(2L, "2"), list(1L, "1"), list(2L, "2")),
+                        schema = c("key", "value"))
```
Comment (HyukjinKwon) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: one more space

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by huaxingao** [comment]
Thanks! @HyukjinKwon @felixcheung

**Pull request closed without merging**